feat:Adds quantity support & prevents zero valued donation tickets

mrsaicharan1 opened the pull request:
<!-- 
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->

<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #3184 

**When organizer tries to adds tickets worth of 0 value**
<img width="1197" alt="Screenshot 2019-06-24 at 11 45 37 PM" src="https://user-images.githubusercontent.com/25197147/60064287-8a33b100-971d-11e9-8d6a-0a4cd059d9f7.png">

**When the ticket buyer tries to add zero valued donation tickets**
<img width="851" alt="Screenshot 2019-06-26 at 7 04 39 AM" src="https://user-images.githubusercontent.com/25197147/60144662-e9a6c500-97e0-11e9-8cb2-4493975d7fdd.png">

#### Short description of what this resolves:
Does not allow donation tickets worth of 0 value to be purchased & adds quantity support

#### Changes proposed in this pull request:

- Added quantity support for donation tickets
- Removed duplicate code segment
- Added check to prevent donation tickets whose value is 0

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes <!-- use `ember test` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

mrsaicharan1 commented:
@CosmicCoder96 @niranjan94 @uds5501 @shreyanshdwivedi  I was able to add support for quantity in donation tickets.
On the other hand, I've added validation checks and added a `min` & `max` attribute to the ticket price . But it's still allowing me to order with 0 price. Could you help me out here?

kushthedude commented:
> @CosmicCoder96 @niranjan94 @uds5501 @shreyanshdwivedi I was able to add support for quantity in donation tickets.
> On the other hand, I've added validation checks and added a `min` & `max` attribute to the ticket price . But it's still allowing me to order with 0 price. Could you help me out here?

Add a server validation for minimum number of tickets . @mrsaicharan1

mrsaicharan1 commented:
> > @CosmicCoder96 @niranjan94 @uds5501 @shreyanshdwivedi I was able to add support for quantity in donation tickets.
> > On the other hand, I've added validation checks and added a `min` & `max` attribute to the ticket price . But it's still allowing me to order with 0 price. Could you help me out here?
> 
> Add a server validation for minimum number of tickets . @mrsaicharan1

Yeah I figured that part out but now suppose I want to check if the tickets contains a donation ticket with amount given as zero, the following the data being sent to create an order. What do I check with in this data?
```
{
  "data": {
    "attributes": {
      "payment-mode": "paypal",
      "country": "India",
      "status": "pending",
      "amount": "10",
      "order-notes": "example"
    },
    "type": "order",
    "relationships": {
      "attendees": {
        "data": [
          {
            "id": "1",
            "type": "attendee"
          }
        ]
      },
      "event": {
        "data": {
          "id": "1",
          "type": "event"
        }
      }
    }
  }
}
```

mrsaicharan1 commented:
I've added a computed function which would disable the order now button if the ticket price for donations is 0. Could you please check this ? @shreyanshdwivedi @uds5501

uds5501 commented on app/components/public/ticket-list.js:
```diff
@@ -16,7 +16,7 @@ export default Component.extend(FormMixin, {
 
   shouldDisableOrderButton: computed('isUnverified', 'hasTicketsInOrder', function() {
-    return !this.hasTicketsInOrder;
+    return !this.hasTicketsInOrder || !this.donationPriceValid;
```
@mrsaicharan1 add 'donationPriceValid' in the watched properties

mrsaicharan1 replied:
@uds5501 Tried it but no luck :/

pradeepgangwar replied:
This should be `return !(this.hasTicketsInOrder && this.donationPriceValid);`

CosmicCoder96 commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,25 @@ export default Component.extend(FormMixin, {
+    return this.data.filterBy('type', 'donation');
+  }),
+
+  donationPriceValid: computed('donationTickets.@each.orderQuantity', 'donationTickets.@each.price', function() {
```
rename this to `isDonationPriceValid`

CosmicCoder96 commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,25 @@ export default Component.extend(FormMixin, {
     ) > 0;
   }),
 
+  donationTickets: computed(function() {
```
you are not watching the property which you are using here, initially this.data will be unresolved, your computed property probably never gets updated hence nothing works.

CosmicCoder96 commented on app/controllers/events/view/tickets/add-order.js:
```diff
@@ -58,6 +58,9 @@ export default Controller.extend({
       let event = order.get('event');
       order.tickets.forEach(ticket => {
+        if (ticket.type === 'donation' && ticket.price === 0) {
+          this.notify.error(this.l10n.t('Price of a donation ticket needs to be greater than 0'));
```
no, don't randomly insert an error message here, it's not blocking anything either way.

CosmicCoder96 commented on app/models/ticket.js:
```diff
@@ -21,8 +21,8 @@ export default ModelBase.extend({
-  minPrice             : attr('number'),
-  maxPrice             : attr('number'),
+  minPrice             : attr('number', { defaultValue: 1 }),
+  maxPrice             : attr('number', { defaultValue: 10 }),
```
don't give a default 10 for maxPrice, let user specify it. 10 might not make sense in certain currencies.

CosmicCoder96 commented on app/templates/components/public/ticket-list.hbs:
```diff
@@ -47,7 +47,13 @@
-                {{currency-symbol eventCurrency}} {{format-number ticket.price}}
+                {{#if (eq ticket.type 'donation') }}
+                <div class="field">
+                  {{input type='number' name="price" placeholder=(t 'Enter Donation') onChange=(action (mut ticket.price)) min=ticket.minPrice max=ticket.maxPrice }}
```
input has a value, not a mut ?

pradeepgangwar commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,25 @@ export default Component.extend(FormMixin, {
+    return this.tickets.filterBy('type', 'donation');
+  }),
+
+  isDonationPriceValid: computed('donationTickets', 'donationTickets.@each.orderQuantity', 'donationTickets.@each.price', function() {
```
This might not be the most elegant way but found it working through this, also please resolve the proxy objects so that we are not using `.get()` Check this https://github.com/fossasia/open-event-frontend/pull/3078/files#r294945298
```js
donationPriceValid: computed('tickets.@each.orderQuantity', 'tickets.@each.price', function() {
    let returnValue = false;
    this.donationTickets.forEach(donationTicket => {
      if (donationTicket.get('orderQuantity') >= 0) {
        if (donationTicket.get('price') > 0) {
          returnValue = true;
        } else {
          returnValue = false;
        }
      }
    });
    return returnValue;
  }),
```

pradeepgangwar commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,25 @@ export default Component.extend(FormMixin, {
+    }
+    return true;
+  }),
+
```
Add this function to ` app/components/public/ticket-list.js`

```js
changePrice(ticket, price) {
      ticket.set('price', price);
      this.order.set('amount', this.total);
      if (!this.total) {
        this.order.set('amount', 0);
      }
    },
```

pradeepgangwar commented on app/templates/components/public/ticket-list.hbs:
```diff
@@ -47,7 +47,13 @@
-                {{currency-symbol eventCurrency}} {{format-number ticket.price}}
+                {{#if (eq ticket.type 'donation') }}
+                <div class="field">
+                  {{input type='number' name="price" placeholder=(t 'Enter Donation') value=ticket.price min=ticket.minPrice max=ticket.maxPrice }}
```
Change this to this:

```
{{input type='number' name="price" placeholder=(t 'Enter Donation') onChange=(action 'changePrice' ticket ticket.price) min=ticket.minPrice max=ticket.maxPrice value=ticket.price}}
```

mrsaicharan1 commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,24 @@ export default Component.extend(FormMixin, {
+    let returnValue = false;
+    this.donationTickets.forEach(donationTicket => {
+      if (donationTicket.orderQuantity >= 0) {
+        if (donationTicket.price > 0) {
```
I've added `donationTickets` to the watchable properties of the computed property. Do I need to explicitly resolve this again by using `donationTicket.get('orderQuantity')`? Will this be returning a promise here `donationTicket.orderQuantity`? @pradeepgangwar @shreyanshdwivedi @uds5501 @CosmicCoder96

mrsaicharan1 commented:
@uds5501 @shreyanshdwivedi 
<img width="769" alt="Screenshot 2019-06-26 at 7 28 51 AM" src="https://user-images.githubusercontent.com/25197147/60145581-20caa580-97e4-11e9-9a80-7a9899cc4e96.png">
Is there a way to display the error popup over the line, rather than pushing the cell contents upwards?

mrsaicharan1 commented:
I've changed the condition such that the donation price should be greater than the minimum price set for it bye the organizer.

mrsaicharan1 commented:
@mariobehling @CosmicCoder96 Please have a look at this.

CosmicCoder96 commented on app/components/forms/wizard/basic-details-step.js:
```diff
@@ -269,6 +269,40 @@ export default Component.extend(FormMixin, EventWizardMixin, {
+              prompt : this.l10n.t('Invalid number')
+            },
+            {
+              type   : 'integer[1..]',
```
the rule be used here is, hat max price should be greater than min price, read sematic ui validation docs.

niranjan94 commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,24 @@ export default Component.extend(FormMixin, {
     ) > 0;
   }),
 
+  donationTickets: computed('data', function() {
```
Use [`filterBy`](https://api.emberjs.com/ember/3.10/functions/@ember%2Fobject%2Fcomputed/filterBy)

niranjan94 commented on app/components/public/ticket-list.js:
```diff
@@ -40,6 +40,24 @@ export default Component.extend(FormMixin, {
+    return this.data.filterBy('type', 'donation');
+  }),
+
+  isDonationPriceValid: computed('donationTickets', 'donationTickets.@each.orderQuantity', 'donationTickets.@each.price', function() {
```
`'donationTickets'` is not needed here

mrsaicharan1 commented:
@CosmicCoder96 Updated with the min/max validation rule.Please review.

mrsaicharan1 commented:
In the validation rule, 
```
window.$.fn.form.settings.rules.checkMaxMin = () => {
	      let { tickets } = this.tickets;
	      for (let ticket of tickets) {
	        if ((ticket.minPrice <= ticket.maxPrice) && (ticket.minPrice > 0) && (ticket.maxPrice > 0)) {
	          return true;
	        }
	      }
	      return false;
	    };
```
when I use object structuring/destructuring,  I get
`` `
checkMaxMin — basic-details-step.js:124TypeError: undefined is not an object (evaluating 'tickets[Symbol.iterator]')` while I try to enter max/min values when creating new donation tickets.
```

But when I just use `let { tickets } = this.tickets;` . It works fine.

niranjan94 commented on app/components/forms/wizard/basic-details-step.js:
```diff
@@ -121,6 +121,15 @@ export default Component.extend(FormMixin, EventWizardMixin, {
 
   getValidationRules() {
+    window.$.fn.form.settings.rules.checkMaxMin = () => {
+      let { tickets } = this.tickets;
```
Incorrect. 

Should be 

```js
let { tickets } = this;
```

But then, why do you even need to store it in a variable ? You can directly use `this.tickets` everywhere right ?

mrsaicharan1 replied:
@niranjan94 Yes, you're right.
